File: 0016-3Sum-Closest/lc0016.py
# Python solution.

class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        sz = len(nums)
        best_match, min_diff = 0, 2147483647
        for i in range(sz-2):
            if i > 0 and nums[i] == nums[i-1]: continue
            left, right = i+1, sz-1
            while left < right:
                sums = nums[i] + nums[left] + nums[right]
                diff = abs(sums-target)
                if diff == 0: return target
                if diff < min_diff:
                    min_diff = diff
                    best_match = sums
                if sums > target: right -= 1
                else: left += 1
        return best_match

File: 0016-3Sum-Closest/lc0016.js
// JavaScript solution.

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */

var threeSumClosest = function(nums, target) {
    let sz = nums.length;
    if (sz < 3) { return -1; }
    else { nums.sort((a,b) => a-b); }
    let best_match = 0, min_diff = 2147483647;

    for (var i = 0; i < sz-2; i++) {
        if (i > 0 && nums[i] === nums[i-1]) continue;
        let left = i+1, right = sz-1;
        while (left < right) {
            let sum = nums[i] + nums[left] + nums[right];
            let diff = Math.abs(sum-target);
            if (diff === 0) { return target; }
            if (diff < min_diff) {
                min_diff = diff;
                best_match = sum;
            }
            if (sum > target) right--;
            else left++;
        }
    }
    return best_match;
};
